### ISSUE TITLE
GPU autograd broken: .data() calls throughout backward pass fail on CUDA tensors

### ISSUE THREAD

**Issue by ngutten** [dollspace-gay/ferrotorch#8]
More stuff from the Pythia 160M implementation attempt - we had issues with gradients on GPU vs gradients on CPU. Here's the writeup from Opus:

**Summary**

The backward pass (autograd) is completely non-functional on GPU tensors. The root cause is pervasive .data() calls (which return Err(GpuTensorNotAccessible) on CUDA tensors) in three critical areas:

  1. The autograd engine itself (autograd/graph.rs)
  2. Shape-manipulation backward functions (grad_fns/shape.rs)
  3. Linear algebra backward functions (grad_fns/linalg.rs)

GPU forward passes work fine. But calling .backward() on any non-trivial computation graph crashes immediately.

**Reproduction**

```
  ferrotorch_gpu::init_cuda_backend()?;
  let device = Device::Cuda(0);

  // Any computation with residual connections or shared tensors:
  let x = ferrotorch_core::from_slice(&[1.0f32, 2.0, 3.0, 4.0], &[2, 2])?
      .to(device)?;
  // x is consumed twice → gradient accumulation triggers in backward
  let a = x.sum_all()?;
  let b = x.mean_all()?;
  let loss = a.add_t(&b)?;
  loss.backward()?;
  // Error: "cannot access GPU tensor data as CPU slice -- call .cpu() first"
```
Root Cause Analysis

Three layers of breakage:

  1. Autograd engine gradient accumulation (most critical — blocks all GPU training)

```
  autograd/graph.rs lines 147-148:
  let a = existing.data()?;  // fails on GPU
  let b = grad.data()?;      // fails on GPU

  tensor.rs accumulate_grad() lines 278, 285:
  let data = incoming.data()?.to_vec();    // fails on GPU
  let incoming_data = incoming.data()?;    // fails on GPU
```
This code runs whenever a non-leaf tensor has multiple consumers (i.e., any residual connection, any tensor reused in multiple operations). This is extremely common in transformer models and cannot be worked around from user code.

  2. Shape backward functions

  In grad_fns/shape.rs, the following backward implementations call grad_output.data() without checking for GPU:
  - ReshapeBackward (line 74)
  - FlattenBackward (line 142)
  - CatBackward (line 560)
  - SqueezeBackward (line 204)
  - UnsqueezeBackward (line 275)

Additionally, view_reshape() in tensor.rs line 137-138 creates the reshaped tensor with grad_fn: None, requires_grad: false, completely severing the autograd graph for GPU reshapes.

  3. Linalg backward functions

  - MmBackward::backward() calls linalg::transpose() and linalg::mm(), both of which call .data() directly
  - bmm_differentiable() line 466 calls result.data()?.to_vec() on the GPU result

  Suggested Fix

Replace all .data() calls in backward functions with .data_vec() (which transparently handles GPU→CPU transfer), or add explicit .cpu() calls before .data(). The activation backward functions (ReluBackward, SigmoidBackward, SoftmaxBackward, etc.) and LayerNormBackward already follow the correct pattern:

```
  // Correct pattern (already used in SoftmaxBackward, LayerNormBackward, etc.):
  let cpu_grad = if grad_output.is_cuda() { grad_output.cpu()? } else { grad_output.clone() };
  let data = cpu_grad.data()?;
```
  The autograd engine accumulation should use .data_vec():
```
  let a = existing.data_vec()?;
  let b = grad.data_vec()?;
```
Or better yet, use the GPU-accelerated add op for accumulation when both tensors are on GPU.

**Impact**

  - GPU forward pass: works
  - GPU backward pass: completely broken
  - Effect: GPU cannot be used for training, only inference
  - Workarounds: Users must wrap every ferrotorch op with custom GPU-safe versions (we implemented ~8 wrappers), but the autograd engine bug is unfixable from user code

**Environment**
  - ferrotorch 0.1.3
  - CUDA 13.0, RTX 3090 Ti
  - Tested with a 12-block Pythia 160M model
